Wooplug 5884/product object cache

prettyboymp opened the pull request:
### Changes proposed in this Pull Request:

<!-- focus: performance [team:platform] -->

This PR introduces a new **Product Object Caching** feature to improve WooCommerce performance by caching product instances during each request, preventing redundant object instantiation overhead when the same product is accessed multiple times within a single page load.

**Key components:**

1. **`ProductCache` class** (`src/Caches/ProductCache.php`):
   - Extends `ObjectCache` to provide product-specific caching
   - Manages product clone modes to preserve meta IDs through WordPress object cache cloning
   - Handles cache/retrieve operations with proper CACHE/DUPLICATE mode switching

2. **`ProductCacheController` class** (`src/Internal/Caches/ProductCacheController.php`):
   - Manages the product caching lifecycle
   - Provides feature flag integration
   - Automatically invalidates cache on product updates (post updates, meta changes, deletions)
   - Sets the `product_objects` cache group as non-persistent

3. **WC_Product_Factory integration**:
   - Checks cache before loading products from the datastore
   - Automatically caches products on load when feature is enabled

4. **Feature flag integration**:
   - Added to `FeaturesController` with proper compatibility declarations
   - Disabled by default (experimental feature)
   - Can be enabled from WooCommerce → Settings → Advanced → Features

**Technical details:**
- Uses WordPress object cache (non-persistent groups) for request-level caching
- Backward compatible - no breaking changes when disabled

Closes [WOOPLUG-5884](https://linear.app/a8c/issue/WOOPLUG-5884/repeated-loading-of-the-same-products-during-a-request-cause) /  #62031 .

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions):

#### Setup

1. Navigate to WooCommerce → Settings → Advanced → Features
2. Enable "Cache Product Objects" feature

#### Test Admin - Product Editing: 

1. Add, Edit, Duplicate different products through the admin.
2. Edit terms, meta, and product content.
3. Verify that changes are saved as expected.  

#### Test Front End: 

1. Go through the shopper flow.
2. Visit a product - preferrebly different types - simple, variations, bundles, subscriptions.
3. Add it to the cart
4. Go through the checkout process.
5. Verify that stocks and sales, if tracked, are updated properly.

### Testing that has already taken place:

- **Unit tests**: Comprehensive test coverage added for `ProductCache` and `ProductCacheController`
- **Integration tests**: Tested cache invalidation triggers (post updates, meta updates, deletions)
- **Backward compatibility**: Verified that disabling the feature maintains existing behavior
- **Clone mode preservation**: Verified meta IDs are preserved through cache but not on user clones
- **Performance**: Tested with product-heavy pages to verify performance improvements
- **WordPress object cache**: Tested with and without persistent object caching (Redis/Memcached)

prettyboymp committed:
Update handling to set clone mode

prettyboymp committed:
add tests

prettyboymp committed:
add more tests, switch to a constant for the feature

prettyboymp committed:
more lint fixes

prettyboymp committed:
lint fixes and logic change around cache clearing for meta changes

prettyboymp committed:
more lint adjustments

prettyboymp committed:
add changelog

prettyboymp commented on plugins/woocommerce/includes/class-wc-product-factory.php:
```diff
@@ -31,6 +32,16 @@ public function get_product( $product_id = false, $deprecated = array() ) {
+		if ( $use_product_cache && empty( $deprecated ) ) {
+			// Nothing should be using the $deprecated argument still, but avoid using cache if they are.
+			$product_cache = wc_get_container()->get( ProductCache::class );
+			$product       = $product_cache->get( $product_id );
```
It won't be the same instance.  The underlying cache runs through WordPress object cache, which clones objects.

prettyboymp committed:
Implement Object Caching for Products

prettyboymp committed:
fix lint issues

prettyboymp commented on plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php:
```diff
@@ -1094,6 +1095,10 @@ protected function update_version_and_type( &$product ) {
 	 * @since 3.0.0
 	 */
 	protected function clear_caches( &$product ) {
+		if ( \Automattic\WooCommerce\Utilities\FeaturesUtil::feature_is_enabled( 'product_instance_caching' ) ) {
```
The ProductCacheController hooks into 'clean_post_cache`, which should handle almost all of these cases.  If any edge cases come up, we can look at those as needed since this is an experimental feature.

prettyboymp commented on plugins/woocommerce/includes/class-wc-product-factory.php:
```diff
@@ -31,6 +32,16 @@ public function get_product( $product_id = false, $deprecated = array() ) {
 		}
 
+		$use_product_cache = \Automattic\WooCommerce\Utilities\FeaturesUtil::feature_is_enabled( 'product_instance_caching' );
+		if ( $use_product_cache && empty( $deprecated ) ) {
```
The current behavior also requires including the deprecated argument every time you want to force a specific product type.  This is a non-issue

prettyboymp commented on plugins/woocommerce/includes/class-wc-product-factory.php:
```diff
@@ -31,6 +32,16 @@ public function get_product( $product_id = false, $deprecated = array() ) {
+			$product_cache = wc_get_container()->get( ProductCache::class );
+			$product       = $product_cache->get( $product_id );
+			if ( $product ) {
+				return $product;
```
I'll need to be sure to include this in the developer advisory - noting that extensions that use the `woocommerce_product_class` will need to return the same result for the life of a request to be compatible.

prettyboymp committed:
Move ProductCache to the internal namespace

prettyboymp committed:
Cast the id before using it for safety.

prettyboymp committed:
Override remove to cast the $id first

prettyboymp committed:
Add @since docs

prettyboymp committed:
Some hardening

prettyboymp commented on plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php:
```diff
@@ -1094,6 +1095,10 @@ protected function update_version_and_type( &$product ) {
 	protected function clear_caches( &$product ) {
+		if ( \Automattic\WooCommerce\Utilities\FeaturesUtil::feature_is_enabled( 'product_instance_caching' ) ) {
+			$product_cache = wc_get_container()->get( ProductCache::class );
+			$product_cache->remove( $product->get_id() );
```
This isn't an issue:

 1. $children / $visible_children (lines 26, 33): These are lazy-loaded - they're null by default and only populated when get_children() is called. Since products are cached right after being read (before these are
  accessed), the cached object has null values, and they get fresh data from the data store when accessed.
  2. Stock levels: Not stored on the parent object. Methods like child_is_in_stock() (line 541) call $this->data_store->child_is_in_stock( $this ) - querying the data store each time.
  3. Price data: get_variation_prices() (line 99) calls $this->data_store->read_price_data() - also queries fresh data each time, not a cached property.

prettyboymp commented on plugins/woocommerce/includes/class-wc-product-factory.php:
```diff
@@ -31,6 +32,16 @@ public function get_product( $product_id = false, $deprecated = array() ) {
+			}
+		}
+
 		$product_type = self::get_product_type( $product_id );
```
This is mentioned above and will need to be noted in the dev advisory.

prettyboymp commented on plugins/woocommerce/src/Internal/Caches/ProductCache.php:
```diff
@@ -0,0 +1,94 @@
+<?php
+declare( strict_types=1 );
```
I've added some explicit conversions.

prettyboymp commented on plugins/woocommerce/src/Internal/Caches/ProductCache.php:
```diff
@@ -0,0 +1,99 @@
+		}
+
+		return null;
+	}
+
+	/**
+	 * Add a product to the cache, or update an already cached product.
+	 *
+	 * Sets the clone mode to CACHE before storing to ensure meta IDs are preserved
+	 * when WordPress object cache clones the object.
+	 *
+	 * @param WC_Product      $product The product to be cached.
+	 * @param int|string|null $id Id of the product to be cached, if null, get_object_id will be used to get it.
+	 * @param int             $expiration Expiration of the cached data in seconds from the current time, or DEFAULT_EXPIRATION to use the default value.
+	 *
+	 * @return bool True on success, false on error.
+	 * @throws \Automattic\WooCommerce\Caching\CacheException Invalid parameter, or null id was passed and get_object_id returns null too.
+	 */
+	public function set( $product, $id = null, int $expiration = self::DEFAULT_EXPIRATION ): bool {
+		if ( null !== $id ) {
+			$id = (int) $id;
+		}
+
+		$original_mode = $product->get_clone_mode();
+		$product->set_clone_mode( \WC_Data::CLONE_MODE_CACHE );
+		$result = parent::set( $product, $id, $expiration );
+		$product->set_clone_mode( $original_mode );
+
+		return $result;
+	}
```
This should be fine as is.

prettyboymp committed:
Add more type conversions to the ProductController

prettyboymp commented on plugins/woocommerce/src/Internal/Caches/ProductCacheController.php:
```diff
@@ -0,0 +1,109 @@
+ * ProductCacheController class file.
+ */
+
+declare( strict_types=1 );
```
Explicit conversions have been added.  All functions/methods called are loosely typed already.

prettyboymp commented on plugins/woocommerce/src/Internal/Caches/ProductCacheController.php:
```diff
@@ -0,0 +1,108 @@
+		add_action( 'before_woocommerce_init', array( $this, 'maybe_set_product_cache_group_as_non_persistent' ) );
+
+		// Handle direct WordPress post updates (bypassing CRUD).
+		add_action( 'clean_post_cache', array( $this, 'maybe_invalidate_product_cache' ), 10, 1 );
```
Because this is in-memory cache, the impact is pretty minimal.  We can look at adding the filter in the future if necessary.

prettyboymp commented on plugins/woocommerce/src/Internal/Caches/ProductCacheController.php:
```diff
@@ -0,0 +1,108 @@
+	 */
+	public function maybe_set_product_cache_group_as_non_persistent() {
+		if ( FeaturesUtil::feature_is_enabled( self::FEATURE_NAME ) ) {
+			wp_cache_add_non_persistent_groups( array( $this->product_cache->get_object_type() ) );
```
The current __sleep()/__wakeup() overrides make persistent caching of WC_Data objects slower than without caching because it causes each object to query and hydrate separately.  So, hopefully no current implementations use this.

Second, this should only affect the new product-object cache group.  Not other existing ones.

prettyboymp committed:
fix lint issues

prettyboymp commented on plugins/woocommerce/includes/class-wc-product-factory.php:
```diff
@@ -31,6 +32,16 @@ public function get_product( $product_id = false, $deprecated = array() ) {
+		if ( $use_product_cache && empty( $deprecated ) ) {
+			// Nothing should be using the $deprecated argument still, but avoid using cache if they are.
+			$product_cache = wc_get_container()->get( ProductCache::class );
+			$product       = $product_cache->get( $product_id );
```
> Extensions that directly call WC_Product_Factory::get_product() and then modify the returned product object will affect subsequent calls to get_product() for the same product ID during the same request. Since the cache returns a clone of the object (via WordPress object cache), modifications made by reference will persist through the cache until invalidation.

This claim doesn't make any sense.  A clone is returned.  The original, cached object, is not modified by modifying extensions modifying the returned object.

jorgeatorres commented on plugins/woocommerce/src/Internal/Caches/ProductCacheController.php:
```diff
@@ -0,0 +1,129 @@
+		$post_id = (int) $post_id;
+		if ( ! FeaturesUtil::feature_is_enabled( self::FEATURE_NAME ) ) {
+			return;
+		}
```
Is this check here just in case something calls the function from the outside? Otherwise, can't we assume that the feature is always enabled _if_ the controller is in use and callbacks hooked?

prettyboymp replied:
I thought it would be needed to deal with tests enabling/disabling the feature, but I was wrong.

jorgeatorres commented on plugins/woocommerce/src/Internal/Caches/ProductCache.php:
```diff
@@ -0,0 +1,124 @@
+	 */
+	protected function validate( $product ): ?array {
+		if ( ! $product instanceof WC_Product ) {
+			return array( 'The supplied product is not an instance of WC_Product, ' . gettype( $product ) );
```
Is `gettype()` here actually useful? For most objects it'll just return "object", which is not very helpful. I don't think we need to provide the class either, the error is probably ok with no information about the specific type or class. What do you think?

prettyboymp replied:
That's good with me.  I was just mimicking the OrderCache response.

jorgeatorres reviewed (commented):
Hey @prettyboymp!

Overall this looks good to me. I guess there might be some edge cases to be expected, but under basic testing everything behaved correctly, so I guess that's fine while experimental.

That said, one thing I noticed on my local setup is that enabling the cache actually resulted in slower execution and a surprising increase in queries. I noticed this on the admin (with Query Monitor) while browsing the **Products** admin script but also reproduced with this simple script in WP-CLI:

   ```php
   <?php
   $start_time = microtime( true );
   
   $products = wc_get_products(
       array(
           'limit' => 100,
       )
   );
   
   var_dump( 'Memory: ' . memory_get_peak_usage( true ) );
   var_dump( 'Time: ' . ( microtime( true ) - $start_time ) );
   var_dump( 'Queries: ' . $GLOBALS['wpdb']->num_queries );
   ```

<img width="711" height="158" alt="Image" src="https://github.com/user-attachments/assets/89aaa7e9-75fd-42f8-a004-3db81e78d79a" />

Maybe something's off with my setup, though I'm not sure what it could be. What are your thoughts?

prettyboymp committed:
Address feedback.
Fix tests that were being skipped.
Clear the Product instance cache after clearing other transients since some call the factory indirectly causing re-caching.
Only load metadata for DUPLICATE clone mode since most cases don't need access to the extended meta.

prettyboymp committed:
Clean post cache after direct deletion of placholder post

prettyboymp committed:
Avoid calls to get_post() during clean_post_cache hook to avoid re-priming cache for code that (incorrectly) depends on the cache being cleaned prior to a db write.

prettyboymp committed:
fix test after removing delete_post hook

prettyboymp commented on plugins/woocommerce/src/Internal/Caches/ProductCacheController.php:
```diff
@@ -0,0 +1,145 @@
+		add_action( 'before_woocommerce_init', array( $this, 'maybe_set_product_cache_group_as_non_persistent' ) );
+
+		// Handle direct WordPress post updates (bypassing CRUD).
+		add_action( 'clean_post_cache', array( $this, 'maybe_invalidate_product_cache_on_clean' ), 10, 2 );
```
The WordPress `cleen_post_cache` hook always uses 2 parameters, even when deleting a post.

prettyboymp committed:
improve teh test_invalidation_respects_feature_flag test

prettyboymp committed:
Merge branch 'trunk' into WOOPLUG-5884/product-object-cache

prettyboymp committed:
Merge branch 'trunk' into WOOPLUG-5884/product-object-cache

prettyboymp committed:
Address PHPStan issues.

jorgeatorres commented on plugins/woocommerce/src/Internal/Caches/ProductCacheController.php:
My concern [last time](https://github.com/woocommerce/woocommerce/pull/62041#pullrequestreview-3545207669) (assuming you can even get to that comment!) was that we were checking if the feature was enabled on every function call.

Now we're not checking at all so all the callbacks are always hooked, even when the feature is disabled.

Is there a way we can check once and hook things only when the feature is actually enabled? I understand this _might_ be a problem for some tests, but maybe those tests can call the initialization method from this controller.

This is not a huge problem, though, but it feels a bit weird that we will add some overhead even when the feature is disabled. What do you think?

jorgeatorres reviewed (changes requested):
Hey @prettyboymp!

I left a tiny comment (which was not fully addressed since the last review) but things are looking good.

The extra queries are no longer happening and performance is pretty much identical when no duplicate loading happens.
I wrote a small test script to check performance and even though it is a big exaggerated, for 50 products loaded ~50 times each, I'm seeing results with product cache that are 4-6 times faster, which is 💯.

prettyboymp committed:
Merge branch 'trunk' into WOOPLUG-5884/product-object-cache

prettyboymp committed:
Only check the feature once before registering hooks instead of each callback.

prettyboymp committed:
Delay the feature check later to avoid early textdomain warnings.

prettyboymp committed:
test cleanup

prettyboymp committed:
fix typo

jorgeatorres reviewed (approved):
LGTM. Thank you @prettyboymp!

Tested the basic flows and ran some benchmarks again using the attached script ([benchmark-product-cache.php.zip](https://github.com/user-attachments/files/24132299/benchmark-product-cache.php.zip)).

prettyboymp committed:
Merge branch 'trunk' into WOOPLUG-5884/product-object-cache

prettyboymp committed:
lint fix after merge conflict
